### PR TITLE
feat: Support multimetric scorer functions

### PR DESCRIPTION
**Pull request by auguste-probabl**
#### Change description

Follow-up to #2898 which made multimetric scorers work only in the `.score()` case.

Multimetric scorers of the form `lambda estimator, X, y: {"metric_1": ..., "metric_2": ..., ...}` are now supported:
```python
from sklearn.datasets import load_iris
from sklearn.metrics import accuracy_score, precision_score, make_scorer
from sklearn.linear_model import LogisticRegression
from skore import evaluate

X, y = load_iris(return_X_y=True)
regressor = LogisticRegression()
report = evaluate(regressor, X, y, splitter=0.2)

def multimetric(y_true, y_pred):
    return {
        "metric_1": accuracy_score(y_true, y_pred),
        "metric_2": precision_score(y_true, y_pred, average=None),
    }

report.metrics.add(make_scorer(multimetric))

report.metrics.summarize().frame()

#                         LogisticRegression
# Metric           Label
# metric_1                          1.000000
# metric_2         0                1.000000
#                  1                1.000000
#                  2                1.000000
# Score                             1.000000
# Accuracy                          1.000000
# Precision        0                1.000000
#                  1                1.000000
#                  2                1.000000
# Recall           0                1.000000
#                  1                1.000000
#                  2                1.000000
# ROC AUC          0                1.000000
#                  1                1.000000
#                  2                1.000000
# Log loss                          0.134784
# Fit time (s)                      0.016429
# Predict time (s)                  0.000191

report.metrics.summarize(metric="multimetric").frame()

#                 LogisticRegression
# Metric   Label
# metric_1                       1.0
# metric_2 0                     1.0
#          1                     1.0
#          2                     1.0
```

This was done by refactoring the Metric internals to be able to output scores in a structured list of rows format, which could then be consumed either in this form by `summarize`, or re-formatted in a human-readable way for the `report.metrics.<metric>` methods.

#### Contribution checklist

Changelog pending

#### AI usage disclosure

AI tools were involved for:
Nothing

### CODE REVIEW

File: skore/src/skore/_sklearn/metrics.py [hunk 1]
```diff
@@ -383,6 +309,230 @@ def new(
                 "Expected a callable, sklearn scorer, or Metric instance."
             )
 
+    def __repr__(self) -> str:
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/metrics.py:312]: moved and made it more robust

File: skore/src/skore/_sklearn/metrics.py [hunk 2]
```diff
@@ -102,6 +135,16 @@ class Metric:
 
     kwargs : dict, default={}
         Default keyword arguments for the scoring function.
+
+    Notes
+    -----
+    A metric's value flows through four layers, from raw to human-readable:
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/metrics.py:141]: since there are now 4 methods that all return the metric score, `__call__` is no longer implemented. Here is how the methods fit together.

File: skore/src/skore/_sklearn/metrics.py [hunk 3]
```diff
@@ -383,6 +309,230 @@ def new(
...
+        return state
+
+    @staticmethod
+    def available(report: EstimatorReport) -> bool:
+        """Whether this metric is applicable to the given report.
+
+        Override this when the metric is not defined for every ML task or estimator
+        (e.g. accuracy is not a regression metric).
+        """
+        return True
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/metrics.py:345]: moved and updated docstrings

**Review by jeromedockes — APPROVED** [review]
LGTM, thank you @auguste-probabl !
